File: AspNetMvcUnleashed/0672329980_source/Chapter21/VB/UnleashedBlog.Tests/Factories/ControllerFactory.vb
Imports UnleashedBlog

''' <summary>
''' Creates one repository and shares the single repository
''' with every controller. When testing, you should retrieve a 
''' controller from this controller factory.
''' </summary>
Public Class ControllerFactory
    Private _blogRepository As BlogRepositoryBase = New FakeBlogRepository()


    Public Function GetBlogController() As BlogController
        Return New BlogController(_blogRepository)
    End Function

    Public Function GetArchiveController() As ArchiveController
        Return New ArchiveController(_blogRepository)
    End Function


    Public Function GetCommentController() As CommentController
        Return New CommentController(_blogRepository)
    End Function


    Public Function GetApplicationController() As ApplicationController
        Return New ApplicationControllerChild(_blogRepository)
    End Function


End Class

Public Class ApplicationControllerChild
    Inherits ApplicationController
    Public Sub New(ByVal blogRepository As BlogRepositoryBase)
        MyBase.New(blogRepository)
    End Sub
End Class

File: AspNetMvcUnleashed/0672329980_source/Chapter21/CS/UnleashedBlog/UnleashedBlog.Tests/Factories/ControllerFactory.cs
using UnleashedBlog.Controllers;
using UnleashedBlog.Models;
using UnleashedBlog.Tests.Models;


namespace UnleashedBlog.Tests.Factories
{
    /// <summary>
    /// Creates one repository and shares the single repository
    /// with every controller. When testing, you should retrieve a 
    /// controller from this controller factory.
    /// </summary>
    public class ControllerFactory
    {
        private BlogRepositoryBase _blogRepository = new FakeBlogRepository();


        public BlogController GetBlogController()
        {
            return new BlogController(_blogRepository);
        }

        public ArchiveController GetArchiveController()
        {
            return new ArchiveController(_blogRepository);
        }


        public CommentController GetCommentController()
        {
            return new CommentController(_blogRepository);
        }


        public ApplicationController GetApplicationController()
        {
            return new ApplicationControllerChild(_blogRepository);
        }


    }

    public class ApplicationControllerChild : ApplicationController
    {
        public ApplicationControllerChild(BlogRepositoryBase blogRepository) 
            : base(blogRepository) { }
    }
}
